### PR TITLE
Implement optimistic mutation results

### PR DESCRIPTION
**Pull request by kitten**

### CODE REVIEW

File: src/operations/write.ts [hunk 1]
```diff
@@ -178,6 +218,9 @@ const writeRoot = (ctx: Context, select: SelectionSet, data: Data) => {
     const fieldValue = data[fieldAlias];
 
     if (ctx.store.updates[fieldName]) {
+      // TODO: Should this really always replace the default logic?
```
Comment (JoviDeCroock) [inline src/operations/write.ts]: I suppose an updater should always have presedence over automatic behavior, `addTodo` --> the consumer of the graphcache decides how this should be written to the store. Neglecting the updater could lead to unexpected behavior

File: src/exchange.ts [hunk 2]
```diff
@@ -37,19 +37,23 @@ const addTypeNames = (op: Operation): Operation => ({
...
 // Returns whether an operation is a query
-const isQueryOperation = (op: Operation): boolean => {
-  return op.operationName === 'query';
-};
+const isQueryOperation = (op: Operation): boolean =>
+  op.operationName === 'query';
+
+// Returns whether an operation is a mutation
+const isMutationOperation = (op: Operation): boolean =>
+  op.operationName === 'query';
```
[the remark this answers is not in the document]
Comment (kitten) [reply src/exchange.ts]: Yep that's just a typo thst I committed since it was temporary code 😅

File: src/operations/write.ts (hunk 1 above)
Comment (kitten) [reply src/operations/write.ts]: I think it'll need to do both at all times. It's literally almost impossible to write a huge fragment for every mutation that has some extra updates 😅

Ultimately all entities will need to be written to the cache for consistency. It's the Query links that need manual updating. I expect that if a GraphQL API with best practices followed returns the data that was mutated by mutations, our cache will already pick up all the changed entities.

The updates job in such a case is just to convey how "Query" links were affected

File: src/operations/query.ts [hunk 3]
```diff
@@ -38,12 +45,13 @@ interface Context {
...
 export const query = (store: Store, request: OperationRequest): QueryResult => {
+  initStoreState(0);
+
   const operation = getMainOperation(request.query);
   const root: Data = Object.create(null);
-
   const result: QueryResult = {
     completeness: 'FULL',
-    dependencies: new Set(),
+    dependencies: getCurrentDependencies(),
```
Comment (JoviDeCroock) [inline src/operations/query.ts]: Oh, I like this 👍

**Pull request merged**